src/votekit/ballot_generator.py

```diff
@@ -1201,4 +1201,144 @@ def __init__(self, **data):
+            
+            for _ in range(num_ballots):
+                # sample ballot type
+                ballot_type = []
+                candidate_count = {b:0 for b in blocs}
+                for _ in range(len(non_zero_cands)):
+                    # choose a bloc as the next slot in the ballot
+                    cand_bloc = list(np.random.choice([bloc, opp_bloc], size =1, 
+                                                      p = ballot_type_distribution))[0]
+                    ballot_type.append(cand_bloc)
+                    candidate_count[cand_bloc] += 1
+
+                    # if there are more candidates to fill
+                    if len(ballot_type) < len(non_zero_cands):
+                        for b in blocs:
+                            # and you are out of non-zero candidates for a slate
+                            if candidate_count[b] == len([c for c in self.slate_to_candidates[b] 
+                                                        if self.pref_interval_by_bloc[bloc][c]>0]):
+                                 # TODO this assumes only two blocs
+                                 o_bloc = set(blocs).difference(b).pop()
+
+                                 # fill out rest of ballot with opposing bloc
+                                 for i in range(len(non_zero_cands) - len(ballot_type)):
+                                     ballot_type.append(o_bloc)
+                    
+                    if len(ballot_type) == len(non_zero_cands):
+                        break
+            
+                cand_ordering_by_bloc = {   }
+                pref_interval = self.pref_interval_by_bloc[bloc]
+
+                for b in blocs:
+                    # create a pref interval dict of only this blocs candidates
+                    bloc_cand_pref_interval = {c:s for c,s in pref_interval.items() 
+                                               if c in self.slate_to_candidates[b] and s>0}
+
+                    cands = list(bloc_cand_pref_interval.keys())
+                    
+                    # if there are no non-zero candidates, skip this bloc
+                    if len(cands) == 0:
+                        continue
+
+                    distribution = [bloc_cand_pref_interval[c] for c in cands]
+                    # normalize
+                    mass = sum(distribution)
+                    distribution = [v/mass for v in distribution]
+
+                    # sample
+                    cand_ordering = np.random.choice(cands, len(cands), 
+                                                     replace=False, p = distribution)
+                    cand_ordering_by_bloc[b] = list(cand_ordering)
+                
+                ranking = []
+                for b in ballot_type:
+                    # append the current first candidate, then remove them from the ordering
+                    ranking.append({cand_ordering_by_bloc[b][0]})
+                    cand_ordering_by_bloc[b].pop(0)
```

Comment:
small thing: pop(0) returns 0th element, you can combine these lines if you want